### PR TITLE
Deploy CLI

### PR DESCRIPTION
**Pull request by bartoszwalacik**

### CODE REVIEW

File: biggerquery/cli.py [hunk 1]
```diff
@@ -130,35 +134,64 @@ def read_project_name_from_setup() -> Optional[str]:
...
-        return 'Project name not found in the project_setup.PROJECT_NAME.'
+        return ''
     else:
         return 'Project name is taken from project_setup.PROJECT_NAME: {0}.'.format(project_name)
 
 
-def find_root_package(project_name: Optional[str], root: Optional[str]) -> Path:
+def find_root_package(project_name: Optional[str], project_dir: Optional[str]) -> Path:
     """
     Finds project package path. Tries first to find locatin in project_setup.PROJECT_NAME,
```
Comment (iwko) [inline biggerquery/cli.py:144]: Tries first to find locati**o**n in project_setup.PROJECT_NAME,

File: biggerquery/cli.py [hunk 2]
```diff
@@ -130,35 +134,64 @@ def read_project_name_from_setup() -> Optional[str]:
...
 def build_project_name_description(project_name: str) -> str:
     if project_name is None:
-        return 'Project name not found in the project_setup.PROJECT_NAME.'
+        return ''
     else:
         return 'Project name is taken from project_setup.PROJECT_NAME: {0}.'.format(project_name)
 
 
-def find_root_package(project_name: Optional[str], root: Optional[str]) -> Path:
+def find_root_package(project_name: Optional[str], project_dir: Optional[str]) -> Path:
```
Comment (beczkowb) [inline biggerquery/cli.py:142]: If you make something optional, then provide default as None

File: biggerquery/cli.py [hunk 3]
```diff
@@ -130,35 +134,64 @@ def read_project_name_from_setup() -> Optional[str]:
...
+        raise ValueError(f'File not found: {file_path.as_posix()}')
+
+    split = file_path.stem.split('-', maxsplit=1)
+    if not len(split) == 2:
+        raise ValueError(f'Invalid file name pattern: {file_path.as_posix()}, expected: *-{{version}}.tar, for example: image-0.1.0.tar')
+    return split[1]
+
+
+def import_deployment_config(deployment_config_path: str):
+    spec = importlib.util.spec_from_file_location("deployment_config", deployment_config_path)
```
Comment (beczkowb) [inline biggerquery/cli.py:171]: Use single quotes or double quotes everywhere possible. In our case single quote.

File: biggerquery/cli.py [hunk 4]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+                             ' If not set, default Config name will be used.'
+                             ' This env name is applied to all biggerquery.Config objects that are defined by'
+                             ' individual workflows as well as to deployment_config.py.')
+
+
+def _add_deploy_parsers_common_arguments(parser):
+    parser.add_argument('-a', '--auth-method',
+                        type=str,
+                        default='local_account',
+                        help="One of two authentication metho: "
```
Comment (iwko) [inline biggerquery/cli.py:269]: help="One of two authentication metho**d**:

File: biggerquery/cli.py [hunk 5]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+                                   description='BiggerQuery CLI deploy-image command -- deploys Docker images to Container Registry.'
+                                   )
+
+    _add_deploy_image_parser_argumentss(parser)
+    _add_deploy_parsers_common_arguments(parser)
+
+
+def _create_deploy_dags_parser(subparsers):
+    parser = subparsers.add_parser('deploy-dags',
+                                   description='BiggerQuery CLI deploy-dags command -- deploys DAG files from local .dagss folder to Composer.')
```
Comment (iwko) [inline biggerquery/cli.py:312]: local **.dags** folder to Composer.

File: biggerquery/cli.py [hunk 6]
```diff
@@ -130,35 +134,64 @@ def read_project_name_from_setup() -> Optional[str]:
...
     else:
-        print('The project_setup.PROJECT_NAME not found. Looking for the --root.')
-        root_module = import_module(root)
+        print(f'The project_setup.PROJECT_NAME not found. Looking for the root module in {project_dir}')
+        root_module = import_module(project_dir)
         return Path(root_module.__file__.replace('__init__.py', ''))
 
 
-def cli_run(root_package: Path,
+def _decode_version_number_from_file_name(file_path: Path):
```
Comment (beczkowb) [inline biggerquery/cli.py:158]: I think we can skip the "private" markers in this module.

File: biggerquery/cli.py [hunk 7]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+    parser.add_argument('-dc', '--deployment-config-path',
+                        type=str,
+                        help="Path to the deployment_config.py file."
+                             " If not set, {current_dir}/deployment_config.py will be used.")
+
+    _add_parsers_common_arguments(parser)
+
+def _create_deploy_parser(subparsers):
+    parser = subparsers.add_parser('deploy',
+                                   description='BiggerQuery CLI deploy command -- performs complete deployment: deploys DAG files to Composer and Docker images to Container Registry.')
```
Comment (iwko) [inline biggerquery/cli.py:294]: Could we remove from description for every command "BiggerQuery CLI {command_name} --command"?  For deploy command example: "Performs complete deployment: deploys DAG files to Composer and Docker images to Container Registry.

File: biggerquery/cli.py [hunk 8]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+
+
+def read_project_package(args):
+    return args.project_package if hasattr(args, 'project_package') else None
+
+
+def _resolve_deployment_config_path(args):
+    if args.deployment_config_path:
+        return args.deployment_config_path
+    return os.getcwd()+'/deployment_config.py'
```
Comment (beczkowb) [inline biggerquery/cli.py:359]: `os.path.join(os.getcwd(), 'deploymen_config.py)`

File: biggerquery/cli.py [hunk 9]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+
+
+def _resolve_dags_dir(args):
+    if args.dags_dir:
+        return args.dags_dir
+    return os.getcwd()+'/.dags'
```
Comment (beczkowb) [inline biggerquery/cli.py:365]: `os.path.join(os.getcwd(), '.dags')`

**Comment by iwko** [comment]
Build fails

File: biggerquery/cli.py [hunk 10]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+                       dags_bucket=_resolve_property(args, 'dags_bucket'),
+                       clear_dags_folder=args.clear_dags_folder,
+                       auth_method=args.auth_method,
+                       vault_endpoint=_resolve_property(args, 'vault_endpoint'),
+                       vault_secret=args.vault_secret,
+                       project_id=_resolve_property(args, 'gcp_project_id')
+                       )
+
+
+def _load_image_from_tar(image_tar_path:str):
```
Comment (beczkowb) [inline biggerquery/cli.py:388]: space

File: biggerquery/deploy.py [hunk 11]
```diff
@@ -32,19 +36,19 @@ def deploy_docker_image(build_ver: str, docker_repository: str, auth_method: str
     return docker_image
 
 
-def deploy_dags_folder(workdir: str, dags_bucket:str, project_id: str, clear_dags_folder=False,
+def deploy_dags_folder(dags_dir: str, dags_bucket:str, project_id: str, clear_dags_folder=False,
```
Comment (beczkowb) [inline biggerquery/deploy.py:39]: space for dags_bucket and typing for clear_dags_folder

File: biggerquery/deploy.py [hunk 12]
```diff
@@ -63,8 +67,8 @@ def blob_URI(blob):
     return f"gs://{blob.bucket.name}/{blob.name}"
 
 
-def upload_DAGs_folder(workdir: str, bucket):
-    dags_dir_path = get_dags_output_dir(workdir)
+def upload_DAGs_folder(dags_dir: str, bucket):
```
Comment (beczkowb) [inline biggerquery/deploy.py:70]: typing for bucket

Comment (beczkowb) [inline biggerquery/deploy.py:70]: maybe dags instead of DAGs?

File: biggerquery/cli.py [hunk 13]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+    group.add_argument('-i', '--image-tar-path',
+                        type=str,
+                        help='Path to a Docker image file. The file name must contain version number with the following naming schema: image-{version}.tar')
+    parser.add_argument('-r', '--docker-repository',
+                        type=str,
+                        help='Name of a local and target Docker repository. Typically, a target repository is hosted by Google Cloud Container Registry.'
+                             ' If so, with the following naming schema: {HOSTNAME}/{PROJECT-ID}/{IMAGE}.'
+                        )
+
+def _add_deploy_dags_parser__arguments(parser):
```
Comment (iwko) [inline biggerquery/cli.py:333]: _add_deploy_dags_parser**_**arguments

File: biggerquery/cli.py [hunk 14]
```diff
@@ -190,36 +238,188 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+
+def _load_image_from_tar(image_tar_path:str):
+    print(f'Loading Docker image from {image_tar_path} ...', )
+
+def _cli_deploy_image(args):
+    if args.image_tar_path:
+        build_ver = _decode_version_number_from_file_name(Path(args.image_tar_path))
+        load_image_from_tar(args.image_tar_path)
+    else:
+        build_ver = build_ver=args.version
```
Comment (iwko) [inline biggerquery/cli.py:396]: double assignment

File: biggerquery/cli.py (hunk 2 above)
Comment (bartoszwalacik) [reply biggerquery/cli.py:142]: it's not mine code, I've only renamed a parameter

File: biggerquery/cli.py (hunk 6 above)
Comment (bartoszwalacik) [reply biggerquery/cli.py:158]: agreed to stay with `_`

File: biggerquery/deploy.py (hunk 12 above)
Comment (bartoszwalacik) [reply biggerquery/deploy.py:70]: Generally, we use CAPS for acronyms in writing as a clear signal that the word is an acronym. What if this rule collide with the Python rule:  "function name should be lowercase" ?

```
get_uri()
```

or

```
get_URI()
```

File: biggerquery/cli.py [hunk 15]
```diff
@@ -190,36 +244,195 @@ def _parse_args(project_name: Optional[str], operations: [str]) -> Namespace:
...
+                        auth_method=args.auth_method,
+                        docker_repository=_resolve_property(args, 'docker_repository'),
+                        vault_endpoint=_resolve_vault_endpoint(args),
+                        vault_secret=args.vault_secret
+                        )
 
 
-def cli() -> None:
+def cli(raw_args) -> None:
+    print("raw_args", raw_args)
```
Comment (iwko) [inline biggerquery/cli.py:421]: do usuniecia

File: test/test_cli.py [hunk 16]
```diff
@@ -242,10 +244,325 @@ def test_should_return_None_when_no_root_property(self):
...
+        f_dev = self._touch_file('image-0.3.0.dev-4b45b638.tar')
+
+        # expect
+        self.assertEqual(_decode_version_number_from_file_name(f_release), '0.1.0')
+        self.assertEqual(_decode_version_number_from_file_name(f_dev), '0.3.0.dev-4b45b638')
+
+        f_release.unlink()
+        f_dev.unlink()
+
+    def test_should_raise_error_when_given_image_file_is_not_tar(self):
```
Comment (iwko) [inline test/test_cli.py:268]: what if file will be .tar.gz?

Comment (bartoszwalacik) [reply test/test_cli.py:268]: won't be :) we genmerate it by ourselves, somewhere in the build process

Comment (iwko) [reply test/test_cli.py:268]: 👍

**Pull request merged**